inst/tests/froll.Rraw

```diff
@@ -801,23 +801,81 @@ test(9999.191, frollmean(c(1:2,NA,4:10), rep(2L,10), adaptive=TRUE, hasNA=FALSE)
 
+### frollsum
+#x = 1:6/2
+#ans1 = frollsum(x, 3)
+#expected = c(rep(NA_real_,2), c(3,4.5,6,7.5))
+#test(9999.201, ans1, expected)
+#n = c(2L, 2L, 3L, 4L, 2L, 3L)
+#ans1 = frollsum(x, n, adaptive=TRUE)
+#expected = c(NA_real_, 1.5, 3, 5, 4.5, 7.5)
+#test(9999.202, ans1, expected)
 
 ## validation
 
 set.seed(108)
-makeNA = function(x, ratio=0.1) {id=sample(length(x), as.integer(length(x) * ratio)); x[id]=NA; x}
-num = 9999.5000 # book 4000 tests for zoo
+makeNA = function(x, ratio=0.1, nf=FALSE) {
+  n = as.integer(length(x) * ratio)
+  id = sample(length(x), n)
+  if (!nf) {
+    x[id] = NA
+  } else {
+    x[id[1:(n/4)]] = NA
+    x[id[(n/4+1):(n/2)]] = NaN
+    x[id[(n/2+1):(3*n/4)]] = -Inf
+    x[id[(3*n/4+1):n]] = +Inf
+  }
+  x
+}
+num = 9999.5000 # book 1000
+## against base to verify exactness of non-finite values, not handled in zoo
+rollfun = function(x, n, FUN, fill=NA_real_, na.rm=FALSE, nf.rm=FALSE) {
+  ans = rep(fill, nx<-length(x))
+  f = match.fun(FUN)
+  if (nf.rm) x[is.infinite(x)] = NA_real_
+  for (i in n:nx) ans[i] = f(x[(i-n+1):i], na.rm=na.rm)
+  ans
+}
+base_compare = function(x, n, funs=c("mean"), algos=c("fast","exact")) {
+  num.step = 0.0001
+  for (fun in funs) {
+    for (na.rm in c(FALSE, TRUE)) {
+      for (fill in c(NA_real_, 0)) {
+        for (algo in algos) {
+          num <<- num + num.step
+          eval(substitute( # so we can have values displayed in output/log rather than variables
+            test(.num,
+                 froll(.fun, x, n, fill=.fill, na.rm=.na.rm, algo=.algo),
+                 rollfun(x, n, FUN=.fun, fill=.fill, na.rm=.na.rm, nf.rm=.nf.rm)),
+            list(.num=num, .fun=fun, .fill=fill, .na.rm=na.rm, .algo=algo, .nf.rm=algo!="exact")
+          ))
+        }
+      }
+    }
+  }
+}
+## random NA non-finite
+x = makeNA(rnorm(1e3), nf=TRUE); n = 50
+base_compare(x, n)
+x = makeNA(rnorm(1e3+1), nf=TRUE); n = 50
+base_compare(x, n)
+x = makeNA(rnorm(1e3), nf=TRUE); n = 51
+base_compare(x, n)
+x = makeNA(rnorm(1e3+1), nf=TRUE); n = 51
+base_compare(x, n)
+num = 9999.6000 # book 3000 tests for zoo
 #### against zoo
 if (requireNamespace("zoo", quietly=TRUE)) {
   drollapply = function(...) as.double(zoo::rollapply(...)) # rollapply is not consistent in data type of answer, force to double
-  zoo_compare = function(x, n) {
+  zoo_compare = function(x, n, funs=c("mean"), algos=c("fast","exact")) {
     num.step = 0.0001
     #### fun, align, na.rm, fill, exact
-    for (fun in c("mean")) { # ,"sum"
+    for (fun in funs) {
       for (align in c("right","center","left")) {
         for (na.rm in c(FALSE, TRUE)) {
           for (fill in c(NA_real_, 0)) {
-            for (algo in c("fast", "exact")) {
+            for (algo in algos) {
+              if (fun=="sum" & algo!="fast") next
```

Comment:
this line shouldn't be here, will amend that when submitting frollsum. It does not have any impact as we don't test for `fun="sum"` yet